log.go

```diff
@@ -164,15 +164,22 @@ func (l *raftLog) findConflict(ents []pb.Entry) uint64 {
 
 // findConflictByTerm takes an (index, term) pair (indicating a conflicting log
 // entry on a leader/follower during an append) and finds the largest index in
-// log l with a term <= `term` and an index <= `index`. If no such index exists
-// in the log, the log's first index is returned.
+// log l with index <= `index` and term <= `term`.
 //
 // The index provided MUST be equal to or less than l.lastIndex(). Invalid
 // inputs log a warning and the input index is returned.
+//
+// If index < firstIndex()-1 then the information about the term at this index
+// has been compacted away. In this case the input index is returned. Similarly,
+// if all terms for entries in [firstIndex()-1, index] are greater than the
+// given term, i.e. the possible suitable entry has been compacted, then
+// firstIndex()-2 is returned. Note that these two cases probably indicate a
```

Comment:
We would return 0 still, because `raftLog.term(0) == 0`. But I refactored this func to guard against `index` underflow, and also made the method comment better.